jetty-server/src/main/java/org/eclipse/jetty/server/AsyncContentProducer.java

```diff
@@ -316,119 +316,144 @@ public boolean isReady()
-                        LOG.debug("using special raw content as transformed content {}", this);
+                        LOG.debug("channel has no new raw content {}", this);
+                    return null;
+                }
+            }
 
-                    // In case the _rawContent was set by consumeAll(), check the httpChannel
-                    // to see if it has a more precise error. Otherwise, the exact same
-                    // special content will be returned by the httpChannel; do not do that
-                    // if the _error flag was set, meaning the current error is definitive.
-                    if (!_error)
-                    {
-                        HttpInput.Content refreshedRawContent = produceRawContent();
-                        if (refreshedRawContent != null)
-                            _rawContent = refreshedRawContent;
-                        _error = _rawContent.getError() != null;
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("refreshed raw content: {} {}", _rawContent, this);
-                    }
+            if (LOG.isDebugEnabled())
+                LOG.debug("transforming raw content {}", this);
+            transformRawContent();
+        }
+    }
 
-                    _transformedContent = _rawContent;
-                    break;
-                }
-                else if (_rawContent.isEmpty())
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("nulling depleted raw content {}", this);
-                    _rawContent.succeeded();
-                    _rawContent = null;
-                    _rawContent = produceRawContent();
-                    if (_rawContent == null)
-                    {
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("produced null raw content, returning null, {}", this);
-                        break;
-                    }
-                }
-                else
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("raw content is not empty {}", this);
-                }
+    private void transformRawContent()
+    {
+        if (_interceptor != null)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("intercepting raw content {}", this);
+            _transformedContent = intercept();
+
+            // If the interceptor generated a special content, _rawContent must become that special content.
+            if (_transformedContent != null && _transformedContent.isSpecial() && _transformedContent != _rawContent)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("interceptor generated a special content, _rawContent must become that special content {}", this);
+                _rawContent.succeeded();
+                _rawContent = _transformedContent;
+                return;
```

Comment:
If the interceptor generated an error content, should you not set `_error` as well?

Reply:
This is already done in the `intercept()` method.